include/nbl/asset/ICPUDescriptorSet.h

```diff
@@ -49,53 +55,61 @@ class NBL_API ICPUDescriptorSet final : public IDescriptorSet<ICPUDescriptorSetL
-            for (uint32_t i = 0u; i <= max_ix; ++i)
-            {
-                auto cloneDescriptor = [](const core::smart_refctd_ptr<IDescriptor>& _desc, uint32_t _depth) -> core::smart_refctd_ptr<IDescriptor> {
-                    if (!_desc)
-                        return nullptr;
-
-                    IAsset* asset = nullptr;
-                    switch (_desc->getTypeCategory())
-                    {
-                    case IDescriptor::EC_BUFFER:
-                        asset = static_cast<ICPUBuffer*>(_desc.get()); break;
-                    case IDescriptor::EC_BUFFER_VIEW:
-                        asset = static_cast<ICPUBufferView*>(_desc.get()); break;
-                    case IDescriptor::EC_IMAGE:
-                        asset = static_cast<ICPUImageView*>(_desc.get()); break;
-                    }
-
-                    auto cp = asset->clone(_depth);
-
-                    switch (_desc->getTypeCategory())
-                    {
-                    case IDescriptor::EC_BUFFER:
-                        return core::smart_refctd_ptr_static_cast<ICPUBuffer>(std::move(cp));
-                    case IDescriptor::EC_BUFFER_VIEW:
-                        return core::smart_refctd_ptr_static_cast<ICPUBufferView>(std::move(cp));
-                    case IDescriptor::EC_IMAGE:
-                        return core::smart_refctd_ptr_static_cast<ICPUImageView>(std::move(cp));
-                    }
-                    return nullptr;
-                };
-
-                auto desc = getDescriptors(i);
-                auto cp_desc = cp->getDescriptors(i);
-
-                const E_DESCRIPTOR_TYPE type = getDescriptorsType(i);
-                for (uint32_t d = 0u; d < desc.size(); ++d)
-                {
-                    cp_desc.begin()[d] = desc.begin()[d];
-                    if (_depth > 0u)
-                    {
-                        cp_desc.begin()[d].desc = cloneDescriptor(cp_desc.begin()[d].desc, _depth-1u);
-                        if (cp_desc.begin()[d].image.sampler && type==EDT_COMBINED_IMAGE_SAMPLER)
-                            cp_desc.begin()[d].image.sampler = core::smart_refctd_ptr_static_cast<ICPUSampler>(cp_desc.begin()[d].image.sampler->clone(_depth-1u));
-                    }
-                }
-            }
+			auto cloneDescriptor = [](const core::smart_refctd_ptr<IDescriptor>& _desc, uint32_t _depth) -> core::smart_refctd_ptr<IDescriptor>
+			{
+				if (!_desc)
+					return nullptr;
+
+				IAsset* asset = nullptr;
+				switch (_desc->getTypeCategory())
+				{
+				case IDescriptor::EC_BUFFER:
+					asset = static_cast<ICPUBuffer*>(_desc.get()); break;
+				case IDescriptor::EC_BUFFER_VIEW:
+					asset = static_cast<ICPUBufferView*>(_desc.get()); break;
+				case IDescriptor::EC_IMAGE:
+					asset = static_cast<ICPUImageView*>(_desc.get()); break;
+				}
+
+				auto cp = asset->clone(_depth);
+
+				switch (_desc->getTypeCategory())
```

Comment:
you already know the type category, shouldn't need to derive it (since you know `t`)

Also not sure why the func needs to be a lambda, and not just inlined into the `for` loop

Reply:
> Also not sure why the func needs to be a lambda, and not just inlined into the for loop

Me neither, it was just like this.

Reply:
Resolved.